Headlamp Plugins CLI: Bulk installations support

Faakhir30 opened the pull request:
Refs #2787, first PR towards original propossed solution.

**Scope and Tasks of this PR:**
- [x] Add parsing and validation of yaml config(Used [ajv](https://www.npmjs.com/package/ajv) for now, would update if maintainers suggests so.)
- [x] Add batch installation module (reuses `PluginManager` for installation, URL and version validations ...)
- [x] Install specific versions pinned, else latest
- [x] Add tests for batch installation module
- [x] Update CLI arguments to support both yaml config file files and direct artifact hub source URLs
- [x] Update Docs (document the plugin yml config)
- [x] Add concurrency control for parallel installations
- [x] add the ability to define DAG using dependencies relations in plugins

The plugin-level config option was removed for now, as it might cause unnecessary confusion, though saved the TODO-based implementation that I removed in git stash, can be redone if required.


---
## Testing:
**individual Plugin with URL:**
  `bin/headlamp-plugin.js install https://artifacthub.io/packages/headlamp/test-123/appcatalog_headlamp_plugin --folderName tmp`
**Bulk installations:**
- Create a file plugin-config.yml:
``` yml
plugins:
  - name: test-app-catalog
    source: https://artifacthub.io/packages/headlamp/test-123/appcatalog_headlamp_plugin
    version: 0.0.3
  - name: ai-plugin
    source: https://artifacthub.io/packages/headlamp/test-123/ai_plugin
    version: 0.0.2
  - name: prometheus
    source: https://artifacthub.io/packages/headlamp/test-123/prometheus_headlamp_plugin
    version: 0.0.3
    dependencies:
      - test-app-catalog

installOptions:
  parallel: true
  maxConcurrent: 2
```
- Then try out: `bin/headlamp-plugin.js install --config tmp/plugin-config.yml --folderName tmp` Make sure `tmp` directory is present.

**Unit Tests:**
For `jest` based tests:
  `f@f:~/w/oss/headlamp/plugins/headlamp-plugin$ npx jest plugin-management/*`

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,138 @@
+    if (this.progressCallback) params.push(this.progressCallback);
+    await PluginManager.install(...params);
+    // Apply plugin configuration if provided
+    if (plugin.config) {
```
cc @yolossn any pointers?

Faakhir30 replied:
@knrt10 Could we confirm if there is at least some need for a plugin level `config` in the template like the shared in PRs description `plugin-config.yml`?
If yes, and we are not sure about it, I can leave  `TODO` for it, open an issue to keep track, and the rest of the code could be reviewed.

knrt10 replied:
AFAIK, I don’t think we have a config right now for a plugin level. You can leave a TODO here

yolossn replied:
AFAIK there isn't a way to configure plugins this way. All the plugin configurations are mostly done by the end user in the frontend.

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,125 @@
+const path = require('path');
```
the name of the folder plugin-management seems redundant. It can be something simpler 1 word like `manager` or something like that.

knrt10 reviewed (changes requested):
The `src/` directory we use is based on typescript for better type safety, consistency and better documentation.

So, I would request you to please address this and change the `manager` in ts as well to avoid inconsistency.

knrt10 commented:
Saw the previous files I think it should be fine to keep it in JS.

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,125 @@
+   * @returns {Promise<void>}
+   */
+  async installPlugin(plugin) {
+    // console.log('Installing plugin', { plugin: plugin.name }); // enable for debugging
```
We should have a proper logging setup.

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,125 @@
+    // console.log('Installing plugin', { plugin: plugin.name }); // enable for debugging
+
+    // Install plugin
+    const params = [plugin.source, this.pluginsDir];
```
We should validate plugin source format before attempting installation, silent failures could occur with undefined parameters

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,125 @@
+        properties: {
+          name: { type: 'string' },
+          source: { type: 'string' },
+          version: { type: 'string' }, // Defaults to latest
```
we should have validation for version format, also there is no constraints on plugin names (could have spaces/special chars).

Maybe something like

```yaml
# for name
pattern: '^[a-z0-9-]+$'  // Only allow lowercase, numbers, hyphens

# for version
pattern: '^\\d+\\.\\d+\\.\\d+' // Semver format
```

Faakhir30 replied:
For name, I've also allowed underscore (except at the start and end), because headlamp plugin names are using underscores somewhere IMO. https://artifacthub.io/packages/headlamp/test-123/ai_plugin.

So used `^[a-z0-9][a-z0-9-_]*[a-z0-9-]$`

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,126 @@
+    const results = [];
+    for (const plugin of config.plugins) {
+      try {
+        await this.installPlugin(plugin);
```
The progress callback is passed directly to `PluginManager.install()` without any context about:

Which plugin is being installed, what number plugin this is in the sequence, how many total plugins there are. I think this can be improved

knrt10 commented on plugins/headlamp-plugin/README.md:
```diff
@@ -52,12 +52,32 @@ headlamp-plugin --help
   headlamp-plugin.js list                   List installed plugins.
-  headlamp-plugin.js install [URL]          Install a plugin from the specified
-                                            Artifact Hub URL.
+  headlamp-plugin.js Install                Install plugin(s) from a configuration
```
Any specific reason why the command is kept `Install` and not `install`?

Faakhir30 replied:
changed by mistake only in readme, reverted

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,126 @@
+      this.pluginsDir === '' ? undefined : this.pluginsDir, 
+      this.headlampVersion === '' ? undefined : this.headlampVersion, 
+      this.progressCallback === null ? undefined : this.progressCallback,
+      null, // signal
```
The signal is hardcoded to null. We should add support for cancellation through AbortController, especially useful for bulk installations:

```js
constructor(pluginsDir = '', headlampVersion = '', progressCallback = null, signal = null) {
  this.signal = signal;
}
```

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,126 @@
+    const config = await this.validateConfig(configPath);
+
+    const results = [];
+    for (const plugin of config.plugins) {
```
Currently sequential installation only is supported. We can add an option for parallel installation with concurrency control for faster install and performance

```js
async installFromConfig(configPath, { parallel = false, maxConcurrent = 3 } = {}) {
```

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,126 @@
+  type: 'object',
+  required: ['plugins'],
+  properties: {
+    plugins: {
```
We should also add support for plugin dependencies and installation order:
```js
dependencies: {
  type: 'array',
  items: { type: 'string' }
}
```

knrt10 commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,126 @@
+    return results;
+  }
+
+  async validateConfig(configPath) {
```
Some important validation is missing

- Duplicate plugin name check
- URL format validation
- Version format validation

knrt10 reviewed (commented):
Some suggestions

Faakhir30 commented:
I've implemented the above suggestions, thanks for the initial review.

Some additional things:
- Regarding Plugin sources validation, I've currently set them to `https://artifacthub.io/packages/headlamp/*`, as this validation is being used previously also, reference: https://github.com/headlamp-k8s/headlamp/blob/08e1e298989284ddc91bf4e6e82785c71657f132/plugins/headlamp-plugin/plugin-management/plugin-management.js#L449-L453
- I've updated the template in this PR's description to test concurrency, and dependency DAG type flow.
- Regarding improving logging further, I've used `progressCallback` to its potential current logs based on installations from the template in this PR's description are following, (any further suggestions for logging styling appreciated):
```
f@f:~/w/oss/lfx/headlamp/plugins/headlamp-plugin$ bin/headlamp-plugin.js install --config
 tmp/config/plugin.yml --folderName tmp
Installing plugins from config { configPath: 'tmp/config/plugin.yml' }
Installation chunks: [ [ 'test-app-catalog', 'ai-plugin' ], [ 'prometheus' ] ]
1 of 3 (test-app-catalog): info: Installing plugin test-app-catalog
1 of 3 (test-app-catalog): info: Fetching Plugin Metadata
2 of 3 (ai-plugin): info: Installing plugin ai-plugin
2 of 3 (ai-plugin): info: Fetching Plugin Metadata
2 of 3 (ai-plugin): info: Plugin Metadata Fetched
2 of 3 (ai-plugin): info: Downloading Plugin
1 of 3 (test-app-catalog): info: Plugin Metadata Fetched
1 of 3 (test-app-catalog): info: Downloading Plugin
2 of 3 (ai-plugin): info: Plugin Downloaded
1 of 3 (test-app-catalog): info: Plugin Downloaded
2 of 3 (ai-plugin): error: Checksum mismatch.
1 of 3 (test-app-catalog): info: Extracting Plugin
1 of 3 (test-app-catalog): info: Plugin Extracted
Moved directory from /tmp/headlamp-plugin-temp-WyoyJw/appcatalog_headlamp_plugin to tmp/appcatalog_headlamp_plugin
1 of 3 (test-app-catalog): success: Plugin Installed
1 of 3 (test-app-catalog): success: Plugin installed successfully
3 of 3 (prometheus): info: Installing plugin prometheus
3 of 3 (prometheus): info: Fetching Plugin Metadata
3 of 3 (prometheus): info: Plugin Metadata Fetched
3 of 3 (prometheus): info: Downloading Plugin
3 of 3 (prometheus): info: Plugin Downloaded
3 of 3 (prometheus): info: Extracting Plugin
3 of 3 (prometheus): info: Plugin Extracted
Moved directory from /tmp/headlamp-plugin-temp-tmM1w4/prometheus_headlamp_plugin to tmp/prometheus_headlamp_plugin
3 of 3 (prometheus): success: Plugin Installed
3 of 3 (prometheus): success: Plugin installed successfully
Bulk installation completed { total: 3, failed: 1, skipped: 0, successful: 2 }
f@f:~/w/oss/lfx/headlamp/plugins/headlamp-plugin$
```
- New considered scenarios:
  - should reject duplicate plugin names
  - should reject duplicate plugin sources
  - should detect circular dependencies
  - should skip plugins with failed/skipped dependencies
  - should respect the maxConcurrent setting
  - should not install dependent plugins in parallel, create optimal chunks within maxConcurrency size limit

- Regarding the folder name `plugin-management`, needed confirmation once more from your end cause that was originally `plugin-management`, can/should I update that to `management`?

chkpwd commented:
Will support for Github and other hosting sites be added or is ArtifactHub the defacto right now?

knrt10 reviewed (commented):
Awesome. Let's update the base branch to `lfx-bulk-plugins`. Will merge all work to main in the end

illume commented:
> Awesome. Let's update the base branch to lfx-bulk-plugins. Will merge all work to main in the end

Do you have a reason to do this?

I'm strongly against using a separate long running feature branch. It's better to follow the normal process and prepare what can be merged into a state where it can be merged, and get it into main branch.

illume commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -0,0 +1,364 @@
+    if (parallel) {
+      // Create optimal chunks that respect dependencies
+      const chunks = this._createOptimalChunks(graph, maxConcurrent);
+      console.log('Installation chunks:', chunks);
```
Is this just for debugging? If so remove, otherwise use progressCallback.

illume reviewed (changes requested):
- Please rebase against main.
- It would make things easier to debug later if the dependency was added in its own commit.
- Please move the "Sync dependencies" commit into the one where you add the dependency.
- Can you please use the context "plugins: plugin-management:" for git commit messages. Remember to capitalize the first letter.

illume commented on plugins/headlamp-plugin/plugin-management/plugin-management.e2e.js:
```diff
@@ -41,7 +41,7 @@ assert.strictEqual(pluginExists, false, 'Plugin should not be initially installe
 // Install the plugin
 const pluginURL = 'https://artifacthub.io/packages/headlamp/test-123/prometheus_headlamp_plugin';
-output = runCommand(`node ../bin/headlamp-plugin.js install ${pluginURL}`);
+output = runCommand(`node ../bin/headlamp-plugin.js install --source ${pluginURL}`);
```
Is this a backwards compatibility break? I mean... does it not work how it did before without "`--source`"?

Faakhir30 replied:
Thanks for pointing out, reverted to use `headlamp-plugin install [URL]`.

illume commented on plugins/headlamp-plugin/plugin-management/multi-plugin-management.js:
```diff
@@ -116,6 +116,7 @@ class MultiPluginManager {
       this.pluginsDir === '' ? undefined : this.pluginsDir, 
       this.headlampVersion === '' ? undefined : this.headlampVersion, 
       this.progressCallback === null ? undefined : this.progressCallback,
+      null, // signal
```
This commit "fix param bug" should be merged in with the commit that added this code.

To make the commit logs cleaner, as it's not necessary to have a separate commit for this. Since it hasn't been merged into the main branch yet.

Faakhir30 commented:
@illume rebased with main, squashed a few commits as suggested, updated commit messages, fixed bugs from CI tests, changes implemented.

> Awesome. Let's update the base branch to lfx-bulk-plugins. Will merge all work to main in the end

@knrt10 I would also prefer to merge to `main` for now, and have a release of this npm package as would be easier to test it if using global npm package install for container. Not sure though yet, reference: https://github.com/kubernetes-sigs/headlamp/pull/2983#issuecomment-2855861448

knrt10 reviewed (approved):
+1. Thank you 😊. Would wait on René’s approval as well

knrt10 commented:
/lgtm
/approve

illume reviewed (approved):
🎉 Thanks!
